### ISSUE TITLE
Server clears Metal cache after every request, destroying KV prefix cache and forcing full re-prefill on every turn

### ISSUE THREAD

**Issue by trevorgordon981** [Blaizzy/mlx-vlm#999]
## Description

`mlx_vlm.server` calls `mx.clear_cache()` and `gc.collect()` after every single generation request (both streaming and non-streaming). This wipes the Metal memory cache, which destroys any KV cache state from the previous request. As a result, every turn in a multi-turn conversation pays the full prefill cost from scratch, even when the conversation shares a long common prefix with the previous turn.

For large models with long conversations, this makes multi-turn latency dramatically worse than it needs to be.

## Location in Source

`mlx_vlm/server.py`, three locations:

```python
# Line ~1077 (streaming path)
mx.clear_cache()
print("Stream finished, cleared cache.")

# Line ~1105 (non-streaming path)  
mx.clear_cache()
gc.collect()
print("Generation finished, cleared cache.")

# Line ~1142 (another path)
mx.clear_cache()
```

## Impact

I serve Qwen3.5 397B A17B (4bit) via mlx_vlm on a Mac Studio M3 Ultra 512GB behind an agent framework (Hermes Agent). A typical conversation builds up to 50,000+ tokens of context over time.

With `prefill-step-size 512`, prefilling 58,000 tokens requires ~115 chunks through a 397B model. This takes several minutes. On the second turn, if the KV cache from the first turn were preserved, only the new tokens (the latest user message + assistant prefix) would need prefilling, which would take seconds instead of minutes.

Instead, every single turn pays the full multi-minute prefill cost because `mx.clear_cache()` wipes everything.

Over the course of a conversation, I observed the server log `Generation finished, cleared cache.` over 1,143 times in one session. Each of those clears threw away KV cache state that could have accelerated the next request.

## Expected Behavior

The server should preserve KV cache state between requests by default for multi-turn conversations. The `prompt_cache_state` parameter already exists in the generate call (line ~1101), suggesting the infrastructure for cache reuse is partially there.

## Suggested Fix

1. **Do not call `mx.clear_cache()` after every request.** Instead, only clear when the model is unloaded, when memory pressure requires it, or when the conversation is reset (detected by a shorter message list than the previous request, which blockops-proxy already detects as `SESSION_RESET`).

2. **Add a server flag** like `--preserve-kv-cache` or `--no-clear-cache` that skips the `mx.clear_cache()` call after generation. Users who need aggressive memory reclamation can keep the current behavior; users serving multi-turn conversations can opt into cache persistence.

3. **Implement prompt cache matching** so that when the next request shares a prefix with the previous one (which is always the case in multi-turn chat), the server reuses the cached KV state and only prefills the new tokens.

## Memory Considerations

The concern behind `mx.clear_cache()` is likely memory reclamation. On a 512GB M3 Ultra serving a single 397B model, memory is not the bottleneck since the model itself uses ~200GB and the KV cache for 60k tokens is modest in comparison. For smaller memory configurations, option 2 (a flag) would let users choose the right tradeoff for their hardware.

## Environment

- mlx_vlm version: 0.4.4
- Model: Qwen3.5 397B A17B 4bit
- macOS, Apple Silicon M3 Ultra 512GB
- Single user serving behind a proxy

**Comment by eloe** [comment]
I think this is the right existing tracker for the issue-first prompt-cache discussion.

A couple of details from current `main` that seem relevant:

- `mlx_vlm/generate.py` already has a `PromptCacheState` class plus prefix-reuse logic (`find_prefix_length`, cache trimming, `prompt_cache_state.update(...)`).
- The server/runtime path still clears aggressively enough that server-level multi-turn reuse remains incomplete in practice, which matches the repro in this issue.

Because of that, I’m not opening a separate PR-first follow-up anymore. Instead, I’d suggest using this issue as the place to discuss the server-facing design for prompt-cache persistence.

## Additional design constraints that probably matter here
If prompt caching is exposed or made persistent at the server layer, I think the issue should explicitly cover:

- cache lifetime / invalidation semantics across turns
- cache keying / session isolation so unrelated requests do not share state accidentally
- bounded eviction behavior (for example size-based and/or time-based expiry) so persistence does not become unbounded retention
- whether persistence is default behavior or guarded by an explicit server flag

In other words, I’d treat TTL / eviction as part of the acceptance criteria for solving this well at the server layer, even if the first fix is just “stop clearing the cache after every request.”

Per maintainer guidance on #1043, I’m using issue-first discussion now before proposing any further PRs in this area.

**Comment by adelnobel** [comment]
I came here after figuring out the same thing, the server is basically unusable for multi turn.

**Comment by Blaizzy** [comment]
Thanks guys! 

We are working to making the server even better everyday, 

Could you share a reproducible example, model and version you are running? 

So we debug and fix it

**Comment by autholykos** [comment]
Reproducible example as requested — with a twist: **the failure is model-class-specific on 0.6.3, and current `main` already fixes it.**

**Setup:** Mac Studio M2 Ultra 192GB · macOS 15 · Python 3.14.3 · mlx 0.31.2 · mlx-lm 0.31.3 · comparing pypi **v0.6.3** vs **main @ `de6fa2d`** (same flags, same machine).

**Server** (f16 KV so APC is eligible; block-mode model — every `make_cache()` entry is plain `KVCache`):

```bash
APC_ENABLED=1 python -m mlx_vlm.server \
  --model mlx-community/Mistral-Small-4-119B-2603-4bit \
  --host 127.0.0.1 --port 8001 --max-kv-size 16384 --max-tokens 8192
```

**Probe:** one stable ~5.9k-token system prefix + 3 different short user questions, then the identical full prompt twice, `temperature=0`, reading `usage.prompt_tokens_details.cached_tokens`:

| build | shared-prefix call latency | cached_tokens |
|---|---|---|
| v0.6.3 | 9.7–10.0 s (every call re-prefills) | **0** |
| main @ `de6fa2d` | **0.65–0.83 s** | **5872–5888** |

Outputs are byte-identical between builds on all probes (temp=0), so the fix is fidelity-safe in our testing; decode throughput is also intact.

**The model-specific part:** on the *same* v0.6.3 server, `mlx-community/Qwen2.5-VL-3B-Instruct-4bit` caches fine (`cached_tokens≈5590`, 2.5 s → 0.3 s). So 0.6.3's prefix-cache loss isn't universal — mistral3/mistral4-family models specifically never restore, while some other block-mode models do. Whatever landed between v0.6.3 and `de6fa2d` repairs the mistral4 restore path.

Two asks:
1. A release carrying this would be very welcome — 0.6.3 is effectively multi-turn/shared-prefix-broken for these models, as others here have said.
2. Happy to bisect the exact fixing commit between v0.6.3 and `de6fa2d` if that's useful for release notes.

(Probe script is ~60 lines stdlib, can share if helpful. Note `--kv-bits` still disables APC entirely on both builds — that's #1174.)

**Comment by DXXS** [comment]
i came here after I just encountered ~this also; for some reason if i use mlx_lm to serve the same model, it is currently caching the repeated portions of the prompt, whereas mlx_vlm (0.6.8) is currently reprocessing the whole thing.  (mlx:0.32.0) (mlx_lm:0.31.3)

**Comment by trevorgordon981** [comment]
Reproducible example on `mlx-vlm 0.6.3`. The failure is specifically shared-prefix reuse across requests with different suffixes; exact full-request reuse works.

### Environment

- MacBook Pro, Apple M5 Max, 128 GB
- macOS 26.5.1 (25F80), arm64
- Python 3.12.8
- `mlx 0.31.2`
- `mlx-lm 0.31.3`
- `mlx-vlm 0.6.3`
- Model: local 8-bit Gemma 4 31B (`model_type=gemma4`, `Gemma4ForConditionalGeneration`), no adapter

### Server

```bash
APC_ENABLED=1 python -m mlx_vlm.server \
  --model "$MODEL" \
  --host 127.0.0.1 \
  --port 8001 \
  --max-kv-size 16384 \
  --max-tokens 128 \
  --log-level INFO
```

### Probe

```python
#!/usr/bin/env python3
import json
import os
import time
import urllib.request

URL = "http://127.0.0.1:8001/v1/chat/completions"
MODEL = os.environ["MODEL"]

PREFIX = (
    "You are a precise technical assistant. Treat this as persistent reference "
    "material. MLX prompt caching should avoid recomputing an identical prefix. "
    * 300
)

def probe(question):
    payload = {
        "model": MODEL,
        "messages": [
            {"role": "system", "content": PREFIX},
            {"role": "user", "content": question},
        ],
        "temperature": 0,
        "max_tokens": 16,
    }
    request = urllib.request.Request(
        URL,
        data=json.dumps(payload).encode(),
        headers={"Content-Type": "application/json"},
    )
    started = time.perf_counter()
    with urllib.request.urlopen(request, timeout=300) as response:
        result = json.load(response)
    elapsed = time.perf_counter() - started
    usage = result["usage"]
    cached = usage.get("prompt_tokens_details", {}).get("cached_tokens")
    print(
        f"{elapsed:.3f}s | prompt_tokens={usage['prompt_tokens']} "
        f"| cached_tokens={cached} | {question}"
    )

probe("Summarize the reference in one sentence.")
probe("What should happen to an identical prompt prefix?")
probe("Why does prefix caching improve multi-turn latency?")
probe("Why does prefix caching improve multi-turn latency?")
```

### Result

```text
14.531s | prompt_tokens=7827 | cached_tokens=1    | Summarize the reference in one sentence.
14.848s | prompt_tokens=7828 | cached_tokens=0    | What should happen to an identical prompt prefix?
15.443s | prompt_tokens=7829 | cached_tokens=0    | Why does prefix caching improve multi-turn latency?
 1.409s | prompt_tokens=7829 | cached_tokens=7813 | Why does prefix caching improve multi-turn latency?
```

The second and third requests retain the same approximately 7.8k-token system prefix but receive no prefix-cache hit and pay the full roughly 15-second prefill. The fourth request repeats the complete third request, reuses 7,813 tokens, and falls to 1.409 seconds.

So APC is active and exact-request restoration works for this model class, but the server does not reuse the longest common prefix when the request suffix changes, which is the normal multi-turn case.

**Comment by f1yn** [comment]
This same issue was preventing me from using a recently fixed Qwen 3.6 A3B model. Each turn, after loading in more files, e.c.t - was causing a huge bubbling effect as the context window grew.

I had Composer 2.5 generate a summary of what I ended up changing locally (fixing the issue for me). My take is that, for actually fixing this upstream, the lines responsible for generating the `inputs_embeds` and `attention_mask` need a stable signature, as having them continually cycle between turns is not workable for actual local usage.

> “Stable signature” as in, the APC lookup salt should only change when something meaningful changed (new image, different tenant, different token prefix) - not every time the prompt gets longer.

---

**Cause (for me):** On the server continuous-batching path, [`BatchGenerator._apc_extra_hash`](https://github.com/Blaizzy/mlx-vlm/blob/4f773d89c769eb2f8fde8595fe66ff90b4920bfc/mlx_vlm/generate/ar.py#L2211-L2233) folds full-sequence `inputs_embeds` and `attention_mask` into APC’s `extra_hash`. Those tensors change every turn as the conversation grows, so [`lookup_exact_cache`](https://github.com/Blaizzy/mlx-vlm/blob/4f773d89c769eb2f8fde8595fe66ff90b4920bfc/mlx_vlm/apc.py#L3007-L3012) never matches (`exact_hits: 0`, full prefill each turn) even with `APC_ENABLED=1`.

**Temporary fix (Currently using a Pi harness):** In [`mlx_vlm/generate/ar.py`](https://github.com/Blaizzy/mlx-vlm/blob/4f773d89c769eb2f8fde8595fe66ff90b4920bfc/mlx_vlm/generate/ar.py#L2225-L2230), inside `_apc_extra_hash`’s `media` dict, comment out:

- [line 2228](https://github.com/Blaizzy/mlx-vlm/blob/4f773d89c769eb2f8fde8595fe66ff90b4920bfc/mlx_vlm/generate/ar.py#L2228): `"embeddings": prompt_kwargs.get("inputs_embeds"),`
- [line 2229](https://github.com/Blaizzy/mlx-vlm/blob/4f773d89c769eb2f8fde8595fe66ff90b4920bfc/mlx_vlm/generate/ar.py#L2229): `"masks": prompt_kwargs.get("attention_mask"),`

Restart with `APC_ENABLED=1`. For long contexts I also set `APC_EXACT_CACHE_ENTRIES=32` (default LRU is [`2`](https://github.com/Blaizzy/mlx-vlm/blob/4f773d89c769eb2f8fde8595fe66ff90b4920bfc/mlx_vlm/apc.py#L2861-L2862)).

**Recommendation:** Drop those two keys from APC `extra_hash` on the batch path (and the same two in [`dispatch.py`](https://github.com/Blaizzy/mlx-vlm/blob/4f773d89c769eb2f8fde8595fe66ff90b4920bfc/mlx_vlm/generate/dispatch.py#L899-L900) for parity). Image/audio/video identity is already covered via `image_hash` / media inputs; token prefix is checked separately.
